core/src/main/java/org/apache/calcite/plan/volcano/VolcanoRuleCall.java

```diff
@@ -285,26 +286,38 @@ private void matchRecurse(int solve) {
         final int parentOrdinal = operand.getParent().ordinalInRule;
         final RelNode parentRel = rels[parentOrdinal];
         final List<RelNode> inputs = parentRel.getInputs();
-        if (operand.ordinalInParent < inputs.size()) {
-          final RelSubset subset =
-              (RelSubset) inputs.get(operand.ordinalInParent);
-          if (operand.getMatchedClass() == RelSubset.class) {
-            successors = subset.set.subsets;
-          } else {
-            successors = subset.getRelList();
+        // if the child is unordered, then add all rels in all subsets to the successor list
+        // unordered can match child in any ordinal
+        if (parentOperand.childPolicy.equals(RelOptRuleOperandChildPolicy.UNORDERED)) {
+          List<RelNode> allRelsInAllSubsets = new ArrayList<>();
+          for (RelNode input : inputs) {
+            RelSubset inputSubset = (RelSubset) input;
+            allRelsInAllSubsets.addAll(inputSubset.getRelList());
           }
+          successors = allRelsInAllSubsets;
         } else {
-          // The operand expects parentRel to have a certain number
-          // of inputs and it does not.
-          successors = ImmutableList.of();
+          if (operand.ordinalInParent < inputs.size()) {
```

Comment:
Is this extra nesting level required?
It looks like `else { if (...) ` could be replaced with `else if (...)`.

PS. I would put `if (operand.ordinalInParent >= inputs.size()) {` as the first `if` branch since it contains less code than the alternative, so it would be easier to read.